storage/aptosdb/src/db_debugger/truncation/mod.rs

```diff
@@ -0,0 +1,414 @@
+    stale_state_value_index::StaleStateValueIndexSchema,
+    transaction_accumulator::TransactionAccumulatorSchema,
+    transaction_info::TransactionInfoSchema,
+    version_data::VersionDataSchema,
+    AptosDB, EventStore, StateStore, TransactionStore,
+};
+use anyhow::Result;
+use aptos_config::config::RocksdbConfigs;
+use aptos_jellyfish_merkle::{node_type::NodeKey, StaleNodeIndex};
+use aptos_schemadb::{
+    schema::{Schema, SeekKeyCodec},
+    ReadOptions, SchemaBatch, DB,
+};
+use aptos_types::{proof::position::Position, transaction::Version};
+use claims::{assert_ge, assert_le, assert_lt};
+use clap::Parser;
+use status_line::StatusLine;
+use std::{
+    fmt::{Display, Formatter},
+    path::PathBuf,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+
+#[derive(Parser)]
+#[clap(about = "Delete all data after the provided version.")]
+pub struct Cmd {
+    #[clap(long, parse(from_os_str))]
+    db_dir: PathBuf,
+
+    #[clap(long)]
+    target_version: u64,
+
+    #[clap(long, default_value = "1000")]
+    ledger_db_batch_size: usize,
+}
+
+struct Progress {
+    current_version: AtomicU64,
+    target_version: Version,
+}
+
+impl Progress {
+    pub fn new(target_version: Version) -> Self {
+        Self {
+            current_version: 0.into(),
+            target_version,
+        }
+    }
+
+    pub fn set_current_version(&self, current_version: Version) {
+        self.current_version
+            .store(current_version, Ordering::Relaxed);
+    }
+}
+
+impl Display for Progress {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "current: {}, target: {}",
+            self.current_version.load(Ordering::Relaxed),
+            self.target_version
+        )
+    }
+}
+
+impl Cmd {
+    pub fn run(self) -> Result<()> {
+        let (ledger_db, state_merkle_db, _kv_db) = AptosDB::open_dbs(
+            &self.db_dir,
+            RocksdbConfigs::default(),
+            /*readonly=*/ false,
+        )?;
+
+        // TODO(grao): Handle kv db once we enable it.
+
+        let ledger_db_version = Self::get_current_version_in_ledger_db(&ledger_db)?
+            .expect("Current version of ledger db must exist.");
+        let state_merkle_db_version =
+            Self::get_current_version_in_state_merkle_db(&state_merkle_db)?
+                .expect("Current version of state merkle db must exist.");
+
+        assert_le!(state_merkle_db_version, ledger_db_version);
+        assert_le!(self.target_version, ledger_db_version);
+
+        println!(
+            "ledger_db_version: {}, state_merkle_db_version: {}, target_version: {}",
+            ledger_db_version, state_merkle_db_version, self.target_version,
+        );
+
+        // TODO(grao): We are using a brute force implementation for now. We might be able to make
+        // it faster, since out data is append only.
+        if self.target_version < state_merkle_db_version {
```

Comment:
I would truncate the ledger db first, otherwise if something happens, it can result in a "valid db" where the node will try to catch up a large chunk of history

Reply:
Truncate ledger first might result in a "invalid db" if the version is smaller than state merkle db version.

(we have code in production to prevent a node catching up a large chunk of history)

Reply:
Yeah I mean, as long as your tool is "resumable", that's better, right? 
Or are you just afraid of putting the db in an invalid state and your tool couldn't make it right?

(K. Doesn't matter too much anyway. As long as we do a checkpoint as a backup.)

Reply:
Added a checkpoint as backup. (I'm afraid of putting db in an invalid state.)